experimental/cuda2/pending_queue_actions.c

```diff
@@ -489,26 +644,38 @@ iree_status_t iree_hal_cuda2_pending_queue_actions_issue(
       // If this semaphore has already signaled past the desired value, we can
       // just ignore it.
       uint64_t value = 0;
-      IREE_RETURN_AND_END_ZONE_IF_ERROR(
-          z0, iree_hal_semaphore_query(semaphores[i], &value));
+      iree_status_t status = iree_hal_semaphore_query(semaphores[i], &value);
+      if (!iree_status_is_ok(status)) {
+        iree_slim_mutex_unlock(&actions->action_mutex);
+        IREE_TRACE_ZONE_END(z0);
+        return status;
+      }
       if (value >= values[i]) continue;
 
       // Try to acquire a CUevent from a device wait timepoint. If so, we can
       // use that CUevent to wait on the device. Otherwise, this action is still
       // not ready.
       CUevent event = NULL;
-      IREE_RETURN_AND_END_ZONE_IF_ERROR(
-          z0, iree_hal_cuda2_event_semaphore_acquire_timepoint_device_wait(
-                  semaphores[i], values[i], &event));
-      if (event) {
-        action->events[action->event_count++] = event;
-      } else {
+      status = iree_hal_cuda2_event_semaphore_acquire_timepoint_device_wait(
+          semaphores[i], values[i], &event);
+      if (!iree_status_is_ok(status)) {
+        iree_slim_mutex_unlock(&actions->action_mutex);
+        IREE_TRACE_ZONE_END(z0);
+        return status;
+      }
+      if (!event) {
         // Clear the scratch fields.
-        action->events = NULL;
         action->event_count = 0;
         action->is_pending = true;
         break;
       }
+      if (action->event_count >= IREE_HAL_CUDA_MAX_WAIT_EVENT_COUNT) {
+        iree_slim_mutex_unlock(&actions->action_mutex);
+        IREE_TRACE_ZONE_END(z0);
+        return iree_make_status(IREE_STATUS_OUT_OF_RANGE,
```

Comment:
RESOURCE_EXHAUSTED is what you want (OUT_OF_RANGE is like an invalid argument passed by the user, not an internal error)